book/layout.md

```diff
@@ -508,6 +494,24 @@ hasn't been computed yet, and the browser will have a bug. We'll come
 back to this issue of dependencies [later](invalidation.md), where it
 will become even more important.
 
+`DocumentLayout` needs some layout code too, though since the
+document always starts in the same place it's pretty simple:
+
+``` {.python}
+class DocumentLayout:
+    def layout(self):
+        # ...
+        self.width = WIDTH - 2*HSTEP
+        self.x = HSTEP
+        self.y = VSTEP
+        child.layout()
+        self.height = child.height + 2*VSTEP
```

Comment:
why 2*VSTEP and not VSTEP here?

Reply:
Ah, to put the padding on both sides.

Reply:
But self.y = VSTEP already puts a padding on the top?

Reply:
Right, but the `height` is needed to add the padding at the bottom too.

Reply:
I still don't get it. The document's top is at VSTEP. It has height equal to content height plus 2*VSTEP. Doesn't this result in a VSTEP padding at the top and 2*VSTEP padding at the bottom?

Reply:
Oh, yeah, I get the confusion. This is actually kind of weird!

So here's how these fields get used. The `document_height` field used for scrolling is just the `DocumentLayout`'s `height` field. We don't add on `y` or anything like that. In fact, that's the only use of this `height` field. So in that sense the correct answer is the current value.

That said it's kind of weird. We can say that the `height` field represents the padding box height, not just content height—but then it's weird that the `y` field represents the content position and not the padding box position. Unfortunately, we don't have a concept of padding, though, so the `y` position here really does have to be `VSTEP`.

So I think the best way to make this make sense is to set `self.height` to the content height and add on `2 * VSTEP` in in the scroll clamping computation. What do you think?

Reply:
I like your suggestion in the last sentence.

Reply:
Done.